Add helper method to set IAQ humidity

benthorner opened the pull request:
Most sensors only report relative humidity and temperature, so this
uses the formula in the datasheet [^1] to make calibration easier.

Notes:

- The datasheet doesn't give the origin of the formula, but it seems
to be based on Tetens Equation (to approximate pressure).

- The example in the datasheet is wrong: the output should be 11.48.

- math.exp is available in MicroPython [^2].

[^1]: https://sensirion.com/media/documents/984E0DD5/61644B8B/Sensirion_Gas_Sensors_Datasheet_SGP30.pdf
[^2]: https://docs.micropython.org/en/latest/library/math.html#math.exp

benthorner committed:
Add helper method to set IAQ humidity

Most sensors only report relative humidity and temperature, so this
uses the formula in the datasheet [^1] to make calibration easier.

Notes:

- The datasheet doesn't give the origin of the formula, but it seems
to be based on Tetens Equation (to approximate pressure).

- The example in the datasheet is wrong: the output should be 11.48.

- math.exp is available in MicroPython [^2].

[^1]: https://sensirion.com/media/documents/984E0DD5/61644B8B/Sensirion_Gas_Sensors_Datasheet_SGP30.pdf
[^2]: https://docs.micropython.org/en/latest/library/math.html#math.exp

tekktrik commented on adafruit_sgp30.py:
```diff
@@ -168,6 +169,19 @@ def set_iaq_humidity(self, gramsPM3):  # pylint: disable=invalid-name
+        """
+        Set the humidity in g/m3 for eCo2 and TVOC compensation algorithm. The absolute humidity is calculated from the temperature and relative humidity (as a percentage).
+        """
+        T = celcius
```
I would just use `celcius` instead of assigning it to `T` to save memory, and it will also fix the pylint error for variable names

benthorner replied:
Sure, amended 👍.

tekktrik commented on adafruit_sgp30.py:
```diff
@@ -168,6 +169,19 @@ def set_iaq_humidity(self, gramsPM3):  # pylint: disable=invalid-name
+        Set the humidity in g/m3 for eCo2 and TVOC compensation algorithm. The absolute humidity is calculated from the temperature and relative humidity (as a percentage).
+        """
+        T = celcius
+        RH = relative_humidity
```
I would also just use `relative_humidty` instead of reassigning.

tekktrik commented on adafruit_sgp30.py:
```diff
@@ -168,6 +169,19 @@ def set_iaq_humidity(self, gramsPM3):  # pylint: disable=invalid-name
 
+    def set_iaq_relative_humidity(self, celcius, relative_humidity):
+        """
+        Set the humidity in g/m3 for eCo2 and TVOC compensation algorithm. The absolute humidity is calculated from the temperature and relative humidity (as a percentage).
```
Looks like this docstring should be broken up into multiple lines:

```suggestion
        Set the humidity in g/m3 for eCo2 and TVOC compensation algorithm.
        The absolute humidity is calculated from the temperature and relative humidity
        (as a percentage).
```

benthorner replied:
This surprised me at first, so I checked and Sphinx only makes a new line / paragraph after a double new line, so it's safe to wrap this manually - amended.

tekktrik commented on adafruit_sgp30.py:
```diff
@@ -26,6 +26,7 @@
 """
 import time
 from adafruit_bus_device.i2c_device import I2CDevice
+from math import exp
```
This import should be moved above the `I2CDevice` import

benthorner replied:
Done.

tekktrik commented on adafruit_sgp30.py:
```diff
@@ -168,6 +169,19 @@ def set_iaq_humidity(self, gramsPM3):  # pylint: disable=invalid-name
+        numerator = ((RH / 100) * 6.112) * exp((17.62 * T) / (243.12 + T))
+        denominator = 273.15 + T
+
+        humidity_gramsPM3 = 216.7 * (numerator / denominator)
```
The variable name doesn't sit well with pylint unfortunately... Should be fine with `humidity_grams_pm3` though!

benthorner replied:
Ah interesting - there's a comment disabling it above. I think your suggestion is clear enough so I've used that - that way we don't need to disable the linter in multiple places.

tekktrik reviewed (changes requested):
I don't think I have the sensor to test this, but in the meantime here are the changes per the CI!  Thanks for submitting!

benthorner commented:
@tekktrik thanks for the speedy first look! Somehow I totally missed this repo has a CI pipeline, which I can now see is passing after fixing all the linting issues. 

I was wondering whether to add a test for the new method but thought it would be too big a step for this PR, seeing as the repo doesn't have any tests currently. What do you think?

tekktrik commented:
Testing isn't the norm for CircuitPython libraries, but adding an example usage to a new or existing example file would be great!  Would you mind doing that?

FoamyGuy commented:
Thanks for working on this @benthorner 

Echoing what tekktrik mentioned. We don't have many automated tests for the libraries. But we always do like to include an example that illustrates how to use new functionality when it gets added. 

If you're up for it please go ahead and add a small script inside of `examples/` directory that shows the new functionality you added.

benthorner committed:
Add example of calibrating IAQ for humidity

In response to [^1].

[^1]: https://github.com/adafruit/Adafruit_CircuitPython_SGP30/pull/35#issuecomment-1074128152

benthorner commented:
@tekktrik @FoamyGuy no problem: https://github.com/adafruit/Adafruit_CircuitPython_SGP30/pull/35/commits/519c9087eab16c7b6e82050cea185018866ebb9d.

I thought it went best in the existing example as we're also setting baselines in there. I also double checked the example file works with the new package and my own SGP30 ✅.

benthorner commented:
@tekktrik @FoamyGuy any chance of this going in soon?

tekktrik reviewed (approved):
Thanks for following up!  I think this looks great, thanks for adding the example!
